languages/python/reference/exercise-concepts/robot-simulator.md

```diff
@@ -0,0 +1,62 @@
+
+- higher-order functions:  a function that takes one or more other functions as arguments, _returning_ a function as its retrun value.  The example uses the built-in `property()` as a higher-order function through `@property`.
+
+- `property()` built-in:  a function that returns a property attribute.  When used as a decorator, this transforms the passed-in method into a _getter_ method for read-only attribute with the same name and docstring.
+
+- assignment:  the example uses assignment for all the instance properties and `instructions` dictionary.
+
+- instance properties: This exercise rquires one or more instance properties to persist passed in data.
+
+- mutability: In the example, knowing there are no protected or private properties in python and so conciously mutating `self.x`, `self.y` and `self.compass` through the called instance methods.
+
+- method arguments: the example `__init__` method needs to take `self`, direction, x, and y (coordinates) as arguments.  It also uses `self` and `commands` (a string) for arguments to the `move()` method.
+- default arguments:  Pre-setting a function argument value to protect against it not being passed..  The example uses `direction = NORTH` and `x=0, y=0` to ensure those values for a robot even if they are not passed in initally.
+
+- dictionaries/mapping type: the example uses a dictionary to map paassed in move arguemtns to methods that perform the moves.  The example also uses a dictionary/mapping created by calling `str.maketrans()`.
+
+- bracket notation in context of dictionaries: the example uses passed in move arguments as keys via bracket notation to look up corresponding methods for moving the robot in the _insructions_ dictionary. 
+
+- iteration/iterators/iterating: the example uses a `for loop` to iterate through the passed in command string.
+
+- composition:  adding functionality from a class by incorporating and instance of that class in a class you are creating.  The example cureates a `robot` by instatiating a `compass`, and assiging it to the `self`.compass property of `robot`.
+
+- modulous operator: the example uses the modulus operator to calculate the appropriate _compass_ setting when calling the `right` compass method.
+
+- lists:  the example stores compass direction constants in a `list`
+
+- for loop: the example uses a `for loop` to iterate through the letters of the passed-in `commands` string and looks up the corresponding values in a dictionary, so that the appropriate methods can be called to move the `robot`.
```

Comment:
This seems somewhat redundant with "iteration" above, though this is the better of the two descriptions. I know there's other forms of iteration than for loops, but I'm not sure we can divorce a for loop from iteration.

Reply:
For this example, the two can be combined I think.  I do think there are exercises and cases where _iteration_ or _iterables_ or _iterators_ do need to be called out differently from a `for loop`.  Comprehensions don't always make it obvious that you're iterating through something for instance.  But there is certainly enough for this exercise as it stands.

Combined wording:

> - iteration: the example uses a `for loop` to iterate through the letters of the passed-in `commands` string and looks up the corresponding values in a dictionary, so that the appropriate methods can be called to move the `robot`.